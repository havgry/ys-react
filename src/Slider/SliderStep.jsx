import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

const SliderStep = ({
  children,
  className,
  ...props
}) => {
  const wrapperProps = {
    className: classNames(
      'ys-slider__data-item',
      className,
    ),
    ...props,
  }

  return (
    <span {...wrapperProps}>
      {children}
    </span>
  )
}

SliderStep.propTypes = {
  children: PropTypes.string.isRequired,
  className: PropTypes.string,
}

SliderStep.defaultProps = {
  className: '',
}

export default SliderStep
