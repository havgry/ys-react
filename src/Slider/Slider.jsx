import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import uniqueId from 'lodash.uniqueid'

import SliderStep from './SliderStep'

import componentPropType from '../helpers/componentPropType'

import '../base.css'
import '@youseedk/dna/css/elements/ys-slider.css'

const id = `slider-${uniqueId()}`

const Slider = ({
  value,
  valueMin,
  valueMax,
  label,
  step,
  onChange,
  className,
  children,
  ...props
}) => {
  const wrapperProps = {
    className: classNames(
      'ys-slider',
      className,
    ),
    ...props,
  }

  const handleChange = (event) => {
    onChange(parseInt(event.target.value, 10), event)
  }

  const getSliderLabel = () => {
    if (children) {
      const assistiveLabels = children
        .map(sliderStep => sliderStep.props.children)
        .join(', ')

      return (
        <span className="ys-u-visuallyhidden">
          {`(valgmulighederne er ${assistiveLabels})`}
        </span>
      )
    }

    return null
  }

  return (
    <div {...wrapperProps}>
      <label
        className="ys-slider__label"
        htmlFor={id}
      >
        <span className="ys-slider__label-text">
          {label}
          {getSliderLabel()}
        </span>
        {children
          && (
            <span className="ys-slider__data-set" aria-hidden="true">
              {children}
            </span>
          )
        }
        <span className="ys-slider__control-wrapper">
          <input
            id={id}
            type="range"
            value={value}
            min={valueMin}
            max={valueMax}
            step={step}
            onChange={handleChange}
            className="ys-slider__control"
          />
        </span>
      </label>
    </div>
  )
}

Slider.propTypes = {
  value: PropTypes.number.isRequired,
  valueMin: PropTypes.number.isRequired,
  valueMax: PropTypes.number.isRequired,
  label: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  children: componentPropType(SliderStep, true),
  step: PropTypes.number,
  className: PropTypes.string,
}

Slider.defaultProps = {
  step: 1,
  children: null,
  className: '',
}

Slider.Step = SliderStep

export default Slider
