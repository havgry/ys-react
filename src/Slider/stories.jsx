import React from 'react'
import { storiesOf, forceReRender } from '@storybook/react'

import { text, number } from '@storybook/addon-knobs'
import { decorate } from '@storybook/addon-actions'
import { Store, State } from '@sambego/storybook-state'

import Slider from './Slider'

const store = new Store({
  value: 2,
})

const updateStore = decorate([(value) => {
  store.set({ value: value[0] })
  return value
}])

store.subscribe(forceReRender)

storiesOf('Slider', module)
  .addParameters({ jest: ['Slider'] })
  .add('Default', () => (
    <State store={store}>
      <Slider
        value={number('value', store.get('value'), {}, 'Content')}
        valueMin={number('valueMin', 1, {}, 'Content')}
        valueMax={number('valueMax', 5, {}, 'Content')}
        step={number('step', 1, {}, 'Content')}
        label={text('label', 'Label', 'Content')}
        onChange={updateStore.action('onChange')}
        className={text('className', '', 'Misc')}
      />
    </State>
  ))
