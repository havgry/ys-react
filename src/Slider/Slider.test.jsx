import React from 'react'
import { shallow, mount } from 'enzyme'

import Slider from './Slider'

describe('<Slider>', () => {
  describe('Props', () => {
    it('Throws propType warning if no `value` prop is passed', () => {
      expect(
        () => shallow(
          <Slider
            valueMin={1}
            valueMax={3}
            label="Label"
            onChange={jest.fn()}
          />,
        ),
      ).toThrow()
    })

    it('Throws propType warning if no `valueMin` prop is passed', () => {
      expect(
        () => shallow(
          <Slider
            value={2}
            valueMax={3}
            label="Label"
            onChange={jest.fn()}
          />,
        ),
      ).toThrow()
    })

    it('Throws propType warning if no `valueMax` prop is passed', () => {
      expect(
        () => shallow(
          <Slider
            value={2}
            valueMin={1}
            label="Label"
            onChange={jest.fn()}
          />,
        ),
      ).toThrow()
    })

    it('Throws propType warning if no `label` prop is passed', () => {
      expect(
        () => shallow(
          <Slider
            value={2}
            valueMin={1}
            valueMax={3}
            onChange={jest.fn()}
          />,
        ),
      ).toThrow()
    })

    it('Throws propType warning if no `onChange` prop is passed', () => {
      expect(
        () => shallow(
          <Slider
            value={2}
            valueMin={1}
            valueMax={3}
            label="Label"
          />,
        ),
      ).toThrow()
    })

    it('Has correct defaultProps', () => {
      const {
        className,
        children,
        step,
      } = mount(
        <Slider
          value={2}
          valueMin={1}
          valueMax={3}
          label="Label"
          onChange={jest.fn()}
        />,
      ).props()

      expect(className).toBe('')
      expect(children).toBe(null)
      expect(step).toBe(1)
    })
  })

  describe('Rendering', () => {
    it('Matches label and input with `for` and `id` attributes', () => {
      const component = shallow(
        <Slider
          value={2}
          valueMin={1}
          valueMax={3}
          label="Label"
          onChange={jest.fn()}
        />,
      )

      expect(
        component
          .find('label')
          .prop('htmlFor'),
      ).toBe(
        component
          .find('input')
          .prop('id'),
      )
    })

    it('Passes other generic props to container element', () => {
      expect(
        shallow(
          <Slider
            id="test"
            value={2}
            valueMin={1}
            valueMax={3}
            label="Label"
            onChange={jest.fn()}
          />,
        ).prop('id'),
      ).toBe('test')
    })

    it('Renders a label for each <Slider.Step>', () => {
      expect(
        shallow(
          <Slider
            value={2}
            valueMin={1}
            valueMax={3}
            label="Label"
            onChange={jest.fn()}
          >
            <Slider.Step>Small</Slider.Step>
            <Slider.Step>Medium</Slider.Step>
            <Slider.Step>Large</Slider.Step>
          </Slider>,
        )
          .find('.ys-slider__data-set')
          .children()
          .length,
      ).toBe(3)
    })

    it('Displays assistive label containing step information', () => {
      expect(
        shallow(
          <Slider
            value={2}
            valueMin={1}
            valueMax={3}
            label="Choose size"
            onChange={jest.fn()}
          >
            <Slider.Step>20 GB</Slider.Step>
            <Slider.Step>40 GB</Slider.Step>
            <Slider.Step>80 GB</Slider.Step>
          </Slider>,
        )
          .find('.ys-slider__label-text span')
          .text(),
      ).toBe('(valgmulighederne er 20 GB, 40 GB, 80 GB)')
    })
  })

  describe('Snapshots', () => {
    it('With optional props', () => {
      expect(
        shallow(
          <Slider
            value={20}
            valueMin={0}
            valueMax={100}
            step={10}
            label="Label"
            onChange={jest.fn()}
            className="lizard"
          />,
        ),
      ).toMatchSnapshot()
    })

    it('With steps', () => {
      expect(
        shallow(
          <Slider
            value={1}
            valueMin={1}
            valueMax={4}
            label="Starbucks sizes"
            onChange={jest.fn()}
          >
            <Slider.Step>Tall</Slider.Step>
            <Slider.Step>Grande</Slider.Step>
            <Slider.Step>Venti</Slider.Step>
            <Slider.Step>Trenta</Slider.Step>
          </Slider>,
        ),
      ).toMatchSnapshot()
    })
  })

  describe('Events', () => {
    it('Delegates `onChange` event when input is changed', () => {
      const changeHandler = jest.fn()
      const mockedEvent = { target: { value: '2' } }

      const component = shallow(
        <Slider
          value={2}
          valueMin={1}
          valueMax={3}
          label="Label"
          onChange={changeHandler}
        />,
      )

      component
        .find('input')
        .simulate('change', mockedEvent)

      // Converts `value` to number and passes entire event as second parameter
      expect(changeHandler).toHaveBeenCalledWith(2, mockedEvent)
    })
  })
})
