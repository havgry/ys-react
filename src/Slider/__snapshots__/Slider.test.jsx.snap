// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Slider> Snapshots With optional props 1`] = `
<div
  className="ys-slider lizard"
>
  <label
    className="ys-slider__label"
    htmlFor="slider-1"
  >
    <span
      className="ys-slider__label-text"
    >
      Label
    </span>
    <span
      className="ys-slider__control-wrapper"
    >
      <input
        className="ys-slider__control"
        id="slider-1"
        max={100}
        min={0}
        onChange={[Function]}
        step={10}
        type="range"
        value={20}
      />
    </span>
  </label>
</div>
`;

exports[`<Slider> Snapshots With steps 1`] = `
<div
  className="ys-slider"
>
  <label
    className="ys-slider__label"
    htmlFor="slider-1"
  >
    <span
      className="ys-slider__label-text"
    >
      Starbucks sizes
      <span
        className="ys-u-visuallyhidden"
      >
        (valgmulighederne er Tall, Grande, Venti, Trenta)
      </span>
    </span>
    <span
      aria-hidden="true"
      className="ys-slider__data-set"
    >
      <SliderStep
        className=""
      >
        Tall
      </SliderStep>
      <SliderStep
        className=""
      >
        Grande
      </SliderStep>
      <SliderStep
        className=""
      >
        Venti
      </SliderStep>
      <SliderStep
        className=""
      >
        Trenta
      </SliderStep>
    </span>
    <span
      className="ys-slider__control-wrapper"
    >
      <input
        className="ys-slider__control"
        id="slider-1"
        max={4}
        min={1}
        onChange={[Function]}
        step={1}
        type="range"
        value={1}
      />
    </span>
  </label>
</div>
`;
